File: app/src/main/java/dev/ragnarok/fenrir/push/ChatEntryFetcher.java
package dev.ragnarok.fenrir.push;

import android.content.Context;
import android.graphics.Bitmap;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.domain.Mode;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.model.Owner;
import dev.ragnarok.fenrir.model.Peer;
import dev.ragnarok.fenrir.util.Optional;
import io.reactivex.rxjava3.core.Single;

public class ChatEntryFetcher {

    public static Single<DialogInfo> getRx(Context context, int accountId, int peerId) {
        Context app = context.getApplicationContext();

        switch (Peer.getType(peerId)) {
            case Peer.USER:
            case Peer.GROUP:
                int ownerId = Peer.toOwnerId(peerId);
                return OwnerInfo.getRx(app, accountId, ownerId)
                        .map(info -> {
                            Owner owner = info.getOwner();

                            DialogInfo response = new DialogInfo();
                            response.title = owner.getFullName();
                            response.img = owner.get100photoOrSmaller();
                            response.icon = info.getAvatar();
                            return response;
                        });
            case Peer.CHAT:
                return Repository.INSTANCE.getMessages()
                        .getConversation(accountId, peerId, Mode.ANY).singleOrError()
                        .flatMap(chat -> NotificationUtils.loadRoundedImageRx(app, chat.get100orSmallerAvatar(), R.drawable.ic_group_chat)
                                .map(Optional::wrap)
                                .onErrorReturnItem(Optional.empty())
                                .map(optional -> {
                                    DialogInfo response = new DialogInfo();
                                    response.title = chat.getTitle();
                                    response.img = chat.get100orSmallerAvatar();
                                    response.icon = optional.get();
                                    return response;
                                }));
        }

        throw new UnsupportedOperationException();
    }

    public static class DialogInfo {
        public String title;
        public String img;
        public Bitmap icon;
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/push/ChatEntryFetcher.kt
package dev.ragnarok.fenrir.push

import android.content.Context
import android.graphics.Bitmap
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.domain.Mode
import dev.ragnarok.fenrir.domain.Repository.messages
import dev.ragnarok.fenrir.model.Peer
import dev.ragnarok.fenrir.push.NotificationUtils.loadRoundedImageRx
import dev.ragnarok.fenrir.util.Optional
import dev.ragnarok.fenrir.util.Optional.Companion.empty
import io.reactivex.rxjava3.core.Single

object ChatEntryFetcher {
    fun getRx(context: Context, accountId: Int, peerId: Int): Single<DialogInfo> {
        val app = context.applicationContext
        when (Peer.getType(peerId)) {
            Peer.USER, Peer.GROUP -> {
                val ownerId = Peer.toOwnerId(peerId)
                return OwnerInfo.getRx(app, accountId, ownerId)
                    .map { info ->
                        val owner = info.owner
                        val response = DialogInfo()
                        response.title = owner.fullName
                        response.img = owner.maxSquareAvatar
                        response.icon = info.avatar
                        response
                    }
            }
            Peer.CHAT, Peer.CONTACT -> return messages
                .getConversation(accountId, peerId, Mode.ANY).singleOrError()
                .flatMap { chat ->
                    loadRoundedImageRx(app, chat.imageUrl, R.drawable.ic_group_chat)
                        .map { Optional.wrap(it) }
                        .onErrorReturnItem(empty())
                        .map { optional ->
                            val response = DialogInfo()
                            response.title = chat.getDisplayTitle()
                            response.img = chat.imageUrl
                            response.icon = optional.get()
                            response
                        }
                }
        }
        throw UnsupportedOperationException()
    }

    class DialogInfo {
        var title: String? = null
        var img: String? = null
        var icon: Bitmap? = null
    }
}